[blog] Supercharging our GHA with Automation API (#5363)
pulumi/docs: content/blog/super-charging-our-github-action-with-the-pulumi-automation-api/index.md

---
title: "GitHub Action now with added Pulumi Automation API"

> h1: "Super Charging our GitHub Action with the Pulumi Automation API"

Reviewer: ```suggestion
h1: "Supercharging our GitHub Action with the Pulumi Automation API"
```

authors: 
  - paul-stack

> #  - simen-olsen

Reviewer: Will we have him listed before going live? Odd to have a "guest post" written by Paul Stack :-)

Author: yes we will :) It's a dual post - I'm waiting for some details from him

tags: 
  - GitHub Actions
  - automation api
date: "2021-02-24"
#meta_image: ""
meta_desc: "Simen Olsen teams up with Paul Stack to introduce the new GitHub Action that has been super charged with the Pulumi Automation API"
---

**Guest Article:** <About Simen here>

In this article, we'll take a look at how our [GitHub Action](https://github.com/pulumi/actions) has been re-architected to be built on top of our [Automation API]({{< relref "/blog/automation-api" >}}).

<!--more-->

> The Pulumi GitHub Action was created back in 2018. This action was a [docker based GitHub Action](https://docs.github.com/en/actions/creating-actions/creating-a-docker-container-action) which

Reviewer: ```suggestion
Pulumi was a GitHub Actions launch partner in 2018, at which time we launched our action using the original [Docker-based GitHub Action type](https://docs.github.com/en/actions/creating-actions/creating-a-docker-container-action). This approach
```

> allowed us to be able to package up all the necessary pre-requisities and the Pulumi CLI to allow users to be able to

Reviewer: ```suggestion
allowed us to package up all the necessary pre-requisities and the Pulumi CLI to allow users to be able to
```

run Pulumi commands. As the Pulumi ecosystem has grown, so too did the container that GitHub Actions run.

> The alternative way to create a GitHub Action is to create a [JavaScript based action](https://docs.github.com/en/actions/creating-actions/creating-a-javascript-action).

Reviewer: ```suggestion
Since then, an alternative way to create a GitHub Action has become available, [by simply writing JavaScript](https://docs.github.com/en/actions/creating-actions/creating-a-javascript-action). At Pulumi, and in our community, we love expressing everything we possibly can in code, so this new support got us quite excited about the "better together" possibilities!
```

> In October 2020, Pulumi released the Automation API. This is  a layer of abstraction that sits across the Pulumi CLI.

Reviewer: ```suggestion
[Enter Pulumi's new Automation API](https://www.pulumi.com/blog/automation-api/), a new feature we launched in public preview in October 2020. This new capability lets you do infrastructure as code &mdash; in pure code &mdash; without even needing to run a CLI.
```

> This allows us to be able to programatically interact with Pulumi without the need to run the Pulumi CLI directly. We were

Reviewer: ```suggestion
This sounded like a perfect match for JavaScript Actions. It allows us to programmatically invoke the right Pulumi commadns in response to actions, without needing to awkwardly script a CLI. Instead, we were
```

> able to leverage the Automation API to rethink our GitHub Action

Reviewer: ```suggestion
able to leverage the Automation API to fundamentally rethink our GitHub Action.
```

> ### GitHub Action v2

Reviewer: I _think_ this section is about how it works, right? So most users don't necessarily need to grok it?

```suggestion
### GitHub Action v2 -- Under the Hood
```

    stackName: config.stackName,
    workDir: workDir,
  });

  const onOutput = (msg: string) => {
    .....
  };

  const actions: Record<Commands, () => Promise<string>> = {
    up: () => stack.up({ onOutput }).then((r) => r.stdout),
    refresh: () => stack.refresh({ onOutput }).then((r) => r.stdout),
    destroy: () => stack.destroy({ onOutput }).then((r) => r.stdout),
    preview: async () => {
      const { stdout, stderr } = await stack.preview();
      onOutput(stdout);
      onOutput(stderr);
      return stdout;
    },
  };
  
  const output = await actions[config.command]();

  const outputs = await stack.outputs();

  for (const [outKey, outExport] of Object.entries(outputs)) {
    core.setOutput(outKey, outExport.value);
    if (outExport.secret) {
      core.setSecret(outExport.value);
    }
  }
};
```

It doesn't have a lot of functionality as the power lays in the Automation API itself. The Action is just an enablement
of the Automation API that can be driven from GitHub Actions. The fact that it has no real functionality, means that as
changes to the Automation API happen, your GitHub Action will be able to take advantage of them without a large level of

> work!

Reviewer: ```suggestion
work! And, of course, you can always write your own custom logic using the Automation API like our action does.
```

### Customizing Your Environment

> The new action allows you to be in control of the environment that it runs on. The docker based action bundled specific tooling

Reviewer: ```suggestion
The new action allows you to be in control of the environment that it runs on. The Docker based action bundled specific tooling
```

versions meaning your CI pipeline was at the mercy of the opinionated container. If you want to pin to specific versions

> of NodeJS / Python / Go / DotNetm or even Pulumi, you now have that power!

Reviewer: ```suggestion
of NodeJS / Python / Go / .NET or even Pulumi, you now have that power!
```

      with:
        creds: ${{ secrets.AZURE_CREDENTIALS }}
```

{{< /choosable >}}

{{% choosable cloud gcp %}}

```yaml
      - uses: google-github-actions/setup-gcloud@master
        with:
          service_account_key: ${{ secrets.GCP_KEY }}
          project_id: ${{ env.PROJECT_ID }}
          export_default_credentials: true
```

{{< /choosable >}}

We suggest that any sensitive environment variables be referenced using
[GitHub Secrets](https://developer.github.com/actions/creating-workflows/storing-secrets/),
and consume them using
[the `secrets` attribute](https://developer.github.com/actions/creating-workflows/workflow-configuration-options/#actions-attributes)
on your workflow's action.

### Faster

The GitHub Action is much smaller and no longer needs to download a (very!) large Docker container. In the dockerfile based
action, up to 3 minutes was spent downloading the container. The new action is much faster as it no longer has this
overhead. Let's look at the comparison:

#### Docker based Action (v1)

![docker based action](old-action.png)

#### Automation API Supercharged Action

![Automation API super charged based action](new-action.png)

> We are eager to hear your feedback on how we can improve your experience with this feature. If you haven’t already signed up for our Community Slack, it’s quick and easy! You can join in on conversations you like and get help from other community members, as well as the Pulumi Team. That’s it for now! 👋

Reviewer: Is there a very simple page we can link to for folks who want to use it? For instance, is https://www.pulumi.com/docs/guides/continuous-delivery/github-actions/ the right place? I would assume 10X more folks just want to kick the tires and try it, compared to those who want to know how it works under the hood.

Reviewer 2: 100% agree. Would love to end with a CTA like:

```suggestion
The new Pulumi GitHub Action is ready for you to use. Here's how to get started:

 - Already use the v1 Pulumi GitHub Action? Migrate to the new one with our [migration guide](TODO-link).
 - New to Pulumi and GitHub Actions? Use our [getting started guide](TODO-link) to set up your first Pulumi workflow.
 
We're eager to hear your feedback on this new Action. If you haven’t already signed up for our Community Slack, it’s quick and easy! You can join in on conversations you like and get help from other community members, as well as the Pulumi Team. That’s it for now! 👋
```

Author: way better! Thanks :)

---
title: "Supercharging our GitHub Action with the Automation API"
authors: 
  - paul-stack
#  - simen-olsen
tags: 

>   - GitHub Actions

Reviewer 2: ```suggestion
  - github-actions
```

Looks like we've used the lowercased and dashed tag in the past

  - automation api

> date: "2021-02-24"

Reviewer 2: ```suggestion
date: "2021-03-01"
```

#meta_image: ""

> meta_desc: "Simen Olsen teams up with Paul Stack to talk about the new GitHub Action that has been super charged with the Pulumi Automation API"

Reviewer 2: ```suggestion
meta_desc: "Simen Olsen and Paul Stack talk about the new GitHub Action powered by the Pulumi Automation API"
```

I think these are better shorter (for rendering on mobile/twitter cards/etc.)?

---

**Guest Article:** <About Simen here>

In this article, we'll take a look at how our [GitHub Action](https://github.com/pulumi/actions) has been re-architected to be built on top of our [Automation API]({{< relref "/blog/automation-api" >}}).

<!--more-->

> Pulumi was a GitHub Actions launch partner in 2018, at which time we launched our action using the original [Docker-based GitHub Action type](https://docs.github.com/en/actions/creating-actions/creating-a-docker-container-action).
> This approach allowed us to package up all the necessary pre-requisities and the Pulumi CLI to allow users to
> be able to run Pulumi commands. As the Pulumi ecosystem has grown, so too did the container that GitHub Actions run.

Reviewer 2: ```suggestion
As a GitHub Actions launch partner in 2018, we launched our original GitHub Action, built as a [Docker container action](https://docs.github.com/en/actions/creating-actions/creating-a-docker-container-action).
This Action packaged all the necessary prerequisites along with the Pulumi CLI to allow users to
be able to run Pulumi commands. As the Pulumi ecosystem has grown, so too did the container run in GitHub Actions.
```

> Since then, an alternative way to create a GitHub Action has become available,
> [by simply writing JavaScript](https://docs.github.com/en/actions/creating-actions/creating-a-javascript-action).
> At Pulumi, and in our community, we love expressing everything we possibly can in code, so this new support got us quite
> excited about the "better together" possibilities! [Enter Pulumi's new Automation API](https://www.pulumi.com/blog/automation-api/),
> a new feature we launched in public preview in October 2020. This new capability lets you do infrastructure as code
> &mdash; in pure code &mdash; without even needing to run a CLI. This sounded like a perfect match for JavaScript Actions.
> It allows us to programmatically invoke the right Pulumi commands in response to actions, without needing to awkwardly
> script a CLI. Instead, we were able to leverage the Automation API to fundamentally rethink our GitHub Action.

Reviewer 2: ```suggestion
Since then, GitHub released an alternative way to create a GitHub Action
[by simply writing JavaScript](https://docs.github.com/en/actions/creating-actions/creating-a-javascript-action).
At Pulumi, and in our community, we love expressing everything we possibly can in code, so this new support got us quite
excited about the "better together" possibilities! [Enter Pulumi's new Automation API](https://www.pulumi.com/blog/automation-api/),
a new feature we launched in public preview in October 2020. This new capability lets you do infrastructure as code
&mdash;in pure code&mdash; without even needing to run a CLI. This sounded like a perfect match for a JavaScript Action.
Creating a JavaScript Action allows us to programmatically invoke the right Pulumi commands in response to actions, without needing to awkwardly
script a CLI. Instead, we were able to leverage the Automation API to fundamentally rethink our GitHub Action.
```

> ### GitHub Action v2 -- Under The Hood

Reviewer 2: ```suggestion
### Under the Hood
```

> The new version of the GitHub action is based largely on the following code:

Reviewer 2: ```suggestion
Let's take a look at the new JavaScript-based GitHub Action and see how it uses Automation API to execute Pulumi:
```

  const config = await makeConfig();

  if (environmentVariables.PULUMI_ACCESS_TOKEN !== '') {
    await pulumiCli.run('login');
  } else if (config.cloudUrl) {
    await pulumiCli.run('login', config.cloudUrl);
  }

  const workDir = resolve(
    environmentVariables.GITHUB_WORKSPACE,
    config.workDir,
  );

  const stack = await LocalWorkspace.selectStack({
    stackName: config.stackName,
    workDir: workDir,
  });

  const onOutput = (msg: string) => {
    .....
  };

  const actions: Record<Commands, () => Promise<string>> = {
    up: () => stack.up({ onOutput }).then((r) => r.stdout),
    refresh: () => stack.refresh({ onOutput }).then((r) => r.stdout),
    destroy: () => stack.destroy({ onOutput }).then((r) => r.stdout),
    preview: async () => {
      const { stdout, stderr } = await stack.preview();
      onOutput(stdout);
      onOutput(stderr);
      return stdout;
    },
  };
  
  const output = await actions[config.command]();

  const outputs = await stack.outputs();

  for (const [outKey, outExport] of Object.entries(outputs)) {
    core.setOutput(outKey, outExport.value);
    if (outExport.secret) {
      core.setSecret(outExport.value);
    }
  }
};
```

> It doesn't have a lot of functionality as the power lays in the Automation API itself. The Action is just an enablement
> of the Automation API that can be driven from GitHub Actions. The fact that it has no real functionality, means that as
> changes to the Automation API happen, your GitHub Action will be able to take advantage of them without a large level of
> work! And, of course, you can always write your own custom logic using the Automation API like our action does.

Reviewer 2: ```suggestion
The Action code is clean and concise because the power lies in the Automation API itself. The Action uses the Automation API to drive Pulumi as part of your GitHub Actions workflows. This means that as Pulumi makes
changes to the Automation API, your workflows will be able to take advantage of them as soon as they're released! And, of course, you can always write your own custom logic using the Automation API like our Action does.
```

### Customizing Your Environment

> The new action allows you to be in control of the environment that it runs on. The Docker based action bundled specific tooling
> versions meaning your CI pipeline was at the mercy of the opinionated container. If you want to pin to specific versions
> of NodeJS / Python / Go / .NET or even Pulumi, you now have that power!

Reviewer 2: ```suggestion
The new Action also allows you to be in control of the environment that it runs on. By comparison, the old Docker-based Action bundled specific tool
versions, which left your CI pipeline at the mercy of our opinionated container. Now, if you want to pin to specific versions
of NodeJS / Python / Go / .NET or even Pulumi, you have that power!
```

> You can even include a matrix strategy to test your Pulumi code against different versions of a language runtime and operating:

Reviewer 2: ```suggestion
You can even include a matrix strategy to test your Pulumi code against different language runtime versions and different operating systems:
```

```yaml
name: 'Sample Pulumi NodeJS Application'
on:
  pull_request:

jobs:
  test-workflow:
    runs-on: ${{ matrix.os }}
    name: Testing Pulumi Preview on ${{ matrix.os }} with NodeJS ${{ matrix.node_version }}
    strategy:
      matrix:
        os: [ubuntu-latest, macos-latest, windows-latest]
        node_version: [10.x, 12.x, 14.x]
      fail-fast: false
    steps:
      - uses: actions/checkout@v2
      - uses: actions/setup-node@v1
        with:
          node-version: ${{ matrix.node_version }}
      - name: Install pulumi
        uses: pulumi/action-install-pulumi-cli@v1.0.1
      - run: npm install
      - uses: actions/pulumi@v2
        env:
          PULUMI_CONFIG_PASSPHRASE: not-a-secret
        with:
          command: preview
          cloud-url: file://~
          stack-name: dev
```

### Cloud Interaction

> One of the things that the previous container did on your behalf was that if you had a `GOOGLE_CREDENTIALS` environment
> variable set then it would run the following commands on your behalf:

Reviewer 2: ```suggestion
The old Docker-based Action would look for the `GOOGLE_CREDENTIALS` environment
variable; if your workflow set that variable, the Action would run the following commands on your behalf:
```

```shell
gcloud auth activate-service-account --key-file=$GOOGLE_APPLICATION_CREDENTIALS
gcloud --quiet auth configure-docker $GOOGLE_DOCKER_HOSTNAME_LIST
```

> With this new action, we wanted to defer the authentication of your cloud to be in your hands. So if you want to authenticate
> with AWS / Azure / Google Cloud then we suggest you use their official actions to do so:

Reviewer 2: ```suggestion
With this new action, you're fully in control of your authentication to your cloud providers. To authenticate
with AWS / Azure / Google Cloud, you can now use their official Actions:
```

{{% chooser cloud "aws,azure,gcp" / %}}

{{% choosable cloud aws %}}

```yaml
      - name: Configure AWS Credentials
        uses: aws-actions/configure-aws-credentials@v1
        with:
          aws-access-key-id: ${{ secrets.AWS_ACCESS_KEY_ID }}
          aws-region: ${{ secrets.AWS_REGION }}
          aws-secret-access-key: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
```

{{< /choosable >}}

{{% choosable cloud azure %}}

```yaml
    - uses: azure/login@v1
      with:
        creds: ${{ secrets.AZURE_CREDENTIALS }}
```

{{< /choosable >}}

{{% choosable cloud gcp %}}

```yaml
      - uses: google-github-actions/setup-gcloud@master
        with:
          service_account_key: ${{ secrets.GCP_KEY }}
          project_id: ${{ env.PROJECT_ID }}
          export_default_credentials: true
```

{{< /choosable >}}

> We suggest that any sensitive environment variables be referenced using
> [GitHub Secrets](https://developer.github.com/actions/creating-workflows/storing-secrets/),
> and consume them using
> [the `secrets` attribute](https://developer.github.com/actions/creating-workflows/workflow-configuration-options/#actions-attributes)
> on your workflow's action.

Reviewer 2: ```suggestion
We recommend that you store all sensitive environment variables, like cloud credentials, in the
Actions [secret storage](https://developer.github.com/actions/creating-workflows/storing-secrets/)
and consume them using
[the `secrets` attribute](https://developer.github.com/actions/creating-workflows/workflow-configuration-options/#actions-attributes)
on your workflow's action.
```

### Faster

> The GitHub Action is much smaller and no longer needs to download a (very!) large Docker container. In the dockerfile based
> action, up to 3 minutes was spent downloading the container. The new action is much faster as it no longer has this
> overhead. Let's look at the comparison:

Reviewer 2: ```suggestion
The new GitHub Action is much smaller and no longer needs to download a (very!) large Docker container. In the Docker-based
Action, we commonly saw workflows take up to 3 minutes just to download the container. The new Action is much faster; let's take a look:
```

> #### Docker based Action (v1)

Reviewer 2: ```suggestion
#### Old Docker-based Action (v1)
```

> ![docker based action](old-action.png)

Reviewer 2: ```suggestion
![Screenshot of the run time of the old Docker-based Pulumi GitHub Action](old-action.png)
```

> #### Automation API Supercharged Action

Reviewer 2: ```suggestion
#### New JavaScript-based Action, powered by Automation API (v2)
```

> ![Automation API super charged based action](new-action.png)

Reviewer 2: ```suggestion
![Screenshot of the run time of the new Pulumi GitHub Action](new-action.png)
```
